File: app/src/main/java/com/unitpricecalculator/util/prefs/Prefs.java
package com.unitpricecalculator.util.prefs;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;
import com.google.common.base.Throwables;
import com.unitpricecalculator.inject.ApplicationContext;
import com.unitpricecalculator.json.ObjectMapper;
import dagger.Reusable;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import javax.inject.Inject;

@Reusable
public final class Prefs {

    private final ObjectMapper objectMapper;
    private final SharedPreferences prefs;

    @Inject
    Prefs(@ApplicationContext Context context, ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
        this.prefs = context.getSharedPreferences(context.getPackageName() + "_prefs",
                Context.MODE_MULTI_PROCESS);
    }

    public String getString(String key) {
        return getString(key, null);
    }

    public String getString(String key, String fallback) {
        return prefs.getString(key, fallback);
    }

    public void putString(String key, String string) {
        prefs.edit().putString(key, string).apply();
    }


    public Set<String> getStringSet(String key, Set<String> fallback) {
        return prefs.getStringSet(key, fallback);
    }

    public void putStringSet(String key, Set<String> set) {
        prefs.edit().putStringSet(key, set).apply();
    }

    public <T> List<T> getList(Class<T> clazz, String key) {
        return getList(clazz, key, new ArrayList<T>());
    }

    public <T> List<T> getList(Class<T> clazz, String key, List<T> fallback) {
        Set<String> set = getStringSet(key, null);
        Log.d("Stevie", "trying to get list: " + set);
        if (set == null) {
            return fallback;
        } else {
            try {
                List<T> list = new ArrayList<>();
                for (String s : set) {
                    list.add(objectMapper.fromJson(clazz, s));
                }
                return list;
            } catch (Exception e) {
                throw Throwables.propagate(e);
            }
        }
    }

    public <T> void putList(String key, List<T> list) {
        Set<String> stringSet = new HashSet<>();
        try {
            for (T t : list) {
                stringSet.add(objectMapper.toJson(t));
            }
            putStringSet(key, stringSet);
        } catch (Exception e) {
            throw Throwables.propagate(e);
        }
    }

    public <T> void addToList(Class<T> clazz, String key, T object) {
        List<T> list = getList(clazz, key);
        list.add(object);
        putList(key, list);
    }

    public <T> T getObject(Class<T> clazz, String key) {
        return getObject(clazz, key, null);
    }

    public <T> T getObject(Class<T> clazz, String key, T fallback) {
        return objectMapper.fromJsonOptional(clazz, getString(key)).or(fallback);
    }

    public void putObject(String key, Object object) {
        prefs.edit().putString(key, objectMapper.toJson(object)).apply();
    }

    public boolean getBoolean(String key) {
        return prefs.getBoolean(key, false);
    }

    public void putBoolean(String key, boolean value) {
        prefs.edit().putBoolean(key, value).apply();
    }

    public void remove(String key) {
        prefs.edit().remove(key).apply();
    }
}

File: app/src/main/java/com/unitpricecalculator/util/prefs/Prefs.kt
package com.unitpricecalculator.util.prefs

import android.content.Context
import android.content.SharedPreferences
import com.unitpricecalculator.inject.ApplicationContext
import com.unitpricecalculator.json.ObjectMapper
import dagger.Reusable
import java.util.ArrayList
import java.util.HashSet
import javax.inject.Inject

@Reusable
class Prefs @Inject internal constructor(
  @ApplicationContext context: Context,
  private val objectMapper: ObjectMapper
) {

  private val prefs: SharedPreferences =
    context.getSharedPreferences(context.packageName + "_prefs", Context.MODE_PRIVATE)

  fun getInt(key: String): Int? {
    if (prefs.contains(key)) {
      return prefs.getInt(key, -1)
    } else {
      return null
    }
  }

  fun getInt(key: String, fallback: Int): Int {
    return prefs.getInt(key, fallback)
  }

  fun putInt(key: String, value: Int?) {
    prefs.edit()
      .also {
        if (value == null) {
          it.remove(key)
        } else {
          it.putInt(key, value)
        }
      }.apply()
  }

  fun getDouble(key: String) = if (prefs.contains(key)) getDouble(key, -1.0) else null

  fun getDouble(key: String, fallback: Double): Double {
    return Double.fromBits(prefs.getLong(key, fallback.toRawBits()))
  }

  fun putDouble(key: String, value: Double) {
    prefs.edit().putLong(key, value.toRawBits()).apply()
  }

  @JvmOverloads
  fun getString(key: String, fallback: String? = null): String? {
    return prefs.getString(key, fallback)
  }

  fun getNonNullString(key: String, fallback: String): String {
    return prefs.getString(key, fallback)!!
  }

  fun putString(key: String, string: String) {
    prefs.edit().putString(key, string).apply()
  }


  fun getStringSet(key: String, fallback: Set<String>?): Set<String>? {
    return prefs.getStringSet(key, fallback)
  }

  fun putStringSet(key: String, set: Set<String>) {
    prefs.edit().putStringSet(key, set).apply()
  }

  fun <T> getList(clazz: Class<T>, key: String): List<T> {
    return getList(clazz, key, ArrayList())
  }

  fun <T> getList(clazz: Class<T>, key: String, fallback: List<T>): List<T> {
    val set = getStringSet(key, null)
    return if (set == null) {
      fallback
    } else {
      try {
        val list = ArrayList<T>()
        for (s in set) {
          list.add(objectMapper.fromJson(clazz, s))
        }
        list
      } catch (e: Exception) {
        throw RuntimeException(e)
      }

    }
  }

  fun <T> putList(key: String, list: List<T>) {
    val stringSet = HashSet<String>()
    try {
      for (t in list) {
        stringSet.add(objectMapper.toJson(t))
      }
      putStringSet(key, stringSet)
    } catch (e: Exception) {
      throw RuntimeException(e)
    }

  }

  fun <T> addToList(clazz: Class<T>, key: String, value: T) {
    val list = getList(clazz, key).toMutableList()
    list.add(value)
    putList(key, list)
  }

  fun <T> getObject(clazz: Class<T>, key: String): T {
    return getObject(clazz, key, null)
  }

  fun <T> getObject(clazz: Class<T>, key: String, fallback: T?): T {
    return objectMapper.fromJsonOptional(clazz, getString(key)).or(fallback!!)
  }

  fun putObject(key: String, `object`: Any) {
    prefs.edit().putString(key, objectMapper.toJson(`object`)).apply()
  }

  fun getBoolean(key: String): Boolean {
    return prefs.getBoolean(key, false)
  }

  fun putBoolean(key: String, value: Boolean) {
    prefs.edit().putBoolean(key, value).apply()
  }

  fun remove(key: String) {
    prefs.edit().remove(key).apply()
  }
}